### PR TITLE
Tremblp/maya 112824/pull push 01 to dev

### PR DESCRIPTION
**Pull request by ppt-adsk**
This pull request introduces the edit as Maya data / merge back to USD workflow (a.k.a pull and push, respectively).  This workflow is used to convert USD data to Maya data for editing, then back to USD once editing is done.  It leverages existing USD import support for edit as Maya (pull), and USD export support for merge back to USD (push).  In a similar way to the import and export prim readers and writers, the edit as Maya / merge back to USD uses prim updaters, created on a per type basis, to perform the conversions.

Once a USD hierarchy has been edited as Maya, it appears in the Outliner hierarchy as parented under its original USD parent.  However, in the Maya data model a separate, hidden Maya hierarchy is created to contain the pulled objects in their Maya representation.  Because the paths to these pulled objects are pure Maya paths, all existing Maya editors and tools can operate on pulled data, and they will appear in the Maya selection as normal Maya objects.

### CODE REVIEW

File: lib/mayaUsd/fileio/inMemoryTranslation.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
+import maya.cmds as cmds
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/fileio/inMemoryTranslation.py:1]: We shouldn't need this file anymore. Let's remove it.

File: lib/mayaUsd/fileio/jobs/readJob.cpp [hunk 2]
```diff
@@ -589,6 +589,11 @@ const MDagPath& UsdMaya_ReadJob::GetMayaRootDagPath() const { return mMayaRootDa
 
 double UsdMaya_ReadJob::timeSampleMultiplier() const { return mTimeSampleMultiplier; }
 
+const UsdMayaPrimReaderContext::ObjectRegistry& UsdMaya_ReadJob::GetNewNodeRegistry() const
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:592]: Should it be available in python?

File: lib/mayaUsd/fileio/primUpdater.cpp [hunk 3]
```diff
@@ -16,30 +16,275 @@
...
 PXR_NAMESPACE_OPEN_SCOPE
 
-UsdMayaPrimUpdater::UsdMayaPrimUpdater(const MFnDependencyNode& depNodeFn, const SdfPath& usdPath)
-    : _dagPath(UsdMayaUtil::getDagPath(depNodeFn))
-    , _mayaObject(depNodeFn.object())
-    , _usdPath(usdPath)
-    , _baseDagToUsdPaths(UsdMayaUtil::getDagPathMap(depNodeFn, usdPath))
+namespace {
+// Set name that will be used to hold all pulled objects
+// MOVED
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdater.cpp:63]: If it's moved, why do we need it here still?

File: lib/mayaUsd/fileio/primUpdater.cpp [hunk 4]
```diff
@@ -16,30 +16,275 @@
...
+
+// Metadata key used to store pull information on a prim
+// MOVED
+const TfToken kPullPrimMetadataKey("Maya:Pull:DagPath");
+
+// Metadata key used to store pull information on a DG node
+// MOVED
+const MString kPullDGMetadataKey("Pull_UfePath");
+
+Ufe::Path usdToMaya(const std::string& pathStr)
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdater.cpp:74]: Do we have a utility for it already? Looks like something useful in general.

File: lib/mayaUsd/fileio/primUpdater.cpp [hunk 5]
```diff
@@ -16,30 +16,275 @@
...
+    return false;
+}
+
+/* static */
+bool UsdMayaPrimUpdater::isAnimated(const MDagPath& path)
+{
+    auto isDagPathAnimated = [](const MDagPath& dagPath) {
+        int     upstreamDependencies = -1;
+        MString pyCommand;
+        pyCommand.format(
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdater.cpp:254]: We should convert this to C++ and expose any missing APIs.

File: lib/mayaUsd/fileio/primUpdater.h [hunk 6]
```diff
@@ -49,71 +52,64 @@ class UsdMayaPrimUpdater
         Push = 1 << 0,
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdater.h:52]: Do we still use Push,Pull,Clear anywhere?

File: lib/mayaUsd/fileio/primUpdaterManager.cpp [hunk 7]
```diff
@@ -0,0 +1,1097 @@
...
+        if (_controlingFlag) {
+            *_controlingFlag = false;
+        }
+    }
+
+private:
+    bool* _controlingFlag { nullptr };
+};
+
+MString pullRootName("__mayaUsd__");
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdaterManager.cpp:643]: Let's put it at the top together with other constants. Update the naming convention as well.

File: lib/mayaUsd/fileio/primUpdaterManager.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    static PrimUpdaterManager& GetInstance();
+
+private:
+    PrimUpdaterManager();
+
+    void onProxyContentChanged(const MayaUsdProxyStageObjectsChangedNotice& notice);
+
+    //! Ensure the Dag pull root exists.  This is the child of the Maya world
+    //! node under which all pulled nodes are created.
+    bool FindOrCreatePullRoot();
```
Comment (kxl-adsk) [inline lib/mayaUsd/fileio/primUpdaterManager.h:65]: We have some inconsistency in naming convention here. Upper case vs lower case method names.

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp [hunk 9]
```diff
@@ -34,6 +34,12 @@
 #include <mayaUsd/ufe/UsdUndoReorderCommand.h>
 #endif
 
+#ifdef UFE_V3_FEATURES_AVAILABLE
+#include <mayaUsd/fileio/primUpdater.h>
+
+#include <ufe/pathString.h> // In UFE v2 but only needed for primUpdater.
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp:40]: Is this needed then? We conditionally compile only when in V3.

File: lib/mayaUsd/ufe/PulledObjectHierarchyHandler.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+#include <mayaUsd/base/api.h>
+
+#include <ufe/hierarchyHandler.h>
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+//! \brief Maya run-time hierarchy handler with support for pulled Maya objects.
+/*!
+    TO DOC
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/PulledObjectHierarchyHandler.h:27]: Now would be a good time to update documentation.

File: lib/mayaUsd/ufe/UsdUIUfeObserver.cpp [hunk 11]
```diff
@@ -77,7 +77,13 @@ void UsdUIUfeObserver::operator()(const Ufe::Notification& notification)
                 "mainChannelBox;");
             MStringArray paths;
             if (MGlobal::executeCommand(mainObjListCmd, paths) && (paths.length() > 0)) {
-                auto ufePath = Ufe::PathString::path(paths[0].asChar());
+                // Skip any non-absolute Maya paths (we know non-Maya ufe paths will always
+                // start with |
+                MString firstPath = paths[0];
+                if (firstPath.substringW(0, 0) != "|")
+                    return;
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdUIUfeObserver.cpp:84]: Why do we need to do this?

File: lib/usd/translators/mayaReferenceUpdater.cpp [hunk 12]
```diff
@@ -25,46 +25,86 @@
...
+    return success;
 }
 
 /* virtual */
-void PxrUsdTranslators_MayaReferenceUpdater::Clear(UsdMayaPrimUpdaterContext* context)
+bool PxrUsdTranslators_MayaReferenceUpdater::DiscardEdits(const UsdMayaPrimUpdaterContext& context)
 {
     const MObject& parentNode = GetMayaObject();
-
     UsdMayaTranslatorMayaReference::UnloadMayaReference(parentNode);
```
Comment (kxl-adsk) [inline lib/usd/translators/mayaReferenceUpdater.cpp:105]: This doesn't discard edits on MayaReference.

File: plugin/adsk/scripts/mayaUSDRegisterStrings.mel [hunk 13]
```diff
@@ -50,6 +50,7 @@ global proc mayaUSDRegisterStrings() {
     register("kMenuClear", "Clear");
     register("kMenuRevertToFile", "Revert to File");
     register("kMenuLayerEditor", "USD Layer Editor");
+    register("kContextMenuLayerEditor", "USD Layer Editor...");
```
Comment (kxl-adsk) [inline plugin/adsk/scripts/mayaUSDRegisterStrings.mel:53]: kMenuLayerEditor and kContextMenuLayerEditor look very similar. Just checking that it's not a mistake to have both.

File: test/lib/ufe/testGroupCmd.py [hunk 14]
```diff
@@ -155,6 +155,12 @@ def testUsdGroup(self):
         groupPath = ufe.Path([mayaPathSegment, usdUtils.createUfePathSegment("/Ball_set/Props/newGroup1")])
         self.assertEqual(globalSelection.front(), ufe.Hierarchy.createItem(groupPath))
 
+        # Group object (a.k.a parent) will be added to selection list. This behavior matches the native Maya group command.
+        globalSelection = ufe.GlobalSelection.get()
```
Comment (kxl-adsk) [inline test/lib/ufe/testGroupCmd.py:159]: How is this change related to the rest of the PR?

File: lib/mayaUsd/fileio/primUpdater.h (hunk 6 above)
Comment (yantor3d) [reply lib/mayaUsd/fileio/primUpdater.h:52]: Pull and Clear are not used currently in `dev`, but are integral to #1759 for allowing the mayaUsdProxyShape to support Maya references (see `mayaReferenceUpdater.cpp`) If they're removed by this PR they'll just get added back in later.

**Comment by ppt-adsk** [comment]
Review comments as of 1-Nov-2021 addressed.  Labeled as do not merge, because of a regression with Maya reference prim creation found internally, to be investigated.

File: lib/usd/translators/mayaReferenceUpdater.cpp (hunk 12 above)
Comment (ppt-adsk) [reply lib/usd/translators/mayaReferenceUpdater.cpp:105]: Quite true.  If you feel this needs to be part of this pull request, I can have a look, otherwise we can consider implementing this in the next phase, which will be for Maya reference workflows.

File: lib/mayaUsd/fileio/primUpdater.cpp [hunk 15]
```diff
@@ -16,30 +16,191 @@
...
+            "import maya.cmds as cmds\n"
+            "if cmds.evaluationManager( query=True, invalidate=True ):\n"
+            "  upstream = cmds.evaluationManager(ust='^1s')\n"
+            "  for node in upstream:\n"
+            "    if cmds.nodeType(node) == 'mayaUsdProxyShape':\n"
+            "      countNonProxyShape -= 1\n"
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primUpdater.cpp:176]: I don't see where countNonProxyShape gets initialized so 1 can be subtracted from it.

File: lib/mayaUsd/fileio/primUpdater.h [hunk 16]
```diff
@@ -49,71 +51,56 @@ class UsdMayaPrimUpdater
...
         Pull = 1 << 1,
         Clear = 1 << 2,
+        AutoPull = 1 << 3,
         All = Push | Pull | Clear
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primUpdater.h:55]: IDK if AutoPull should be in All?

File: lib/mayaUsd/fileio/primUpdater.cpp (hunk 15 above)
Comment (ppt-adsk) [reply lib/mayaUsd/fileio/primUpdater.cpp:176]: Absolutely, this will generate a NameError exception.  Will investigate, good catch.

File: lib/mayaUsd/fileio/primUpdater.h (hunk 16 above)
Comment (ppt-adsk) [reply lib/mayaUsd/fileio/primUpdater.h:55]: I don't know either.  I will assume the above is correct, until shown otherwise.

File: lib/usd/translators/mayaReferenceUpdater.cpp [hunk 17]
```diff
@@ -25,46 +25,86 @@
...
+    const SdfPath& srcSdfPath,
+    SdfLayerRefPtr dstLayer,
+    const SdfPath& dstSdfPath)
 {
-    const UsdPrim& usdPrim = GetUsdPrim<MayaUsd_SchemasMayaReference>(*context);
-    const MObject& parentNode = GetMayaObject();
+    bool success = false;
 
-    UsdMayaTranslatorMayaReference::update(usdPrim, parentNode);
+    // We are looking for a very specific configuration in here
```
Comment (kxl-adsk) [inline lib/usd/translators/mayaReferenceUpdater.cpp:71]: Worth mentioning that this code is going to change very soon as it's only a "prototype" code.

**Review by kxl-adsk — APPROVED** [review]
I'm accepting this, but there is still more cleanup to do in subsequent PRs.

**Review by kxl-adsk — DISMISSED** [review]

**Review by kxl-adsk — APPROVED** [review]
